FirebaseSegmentation/Sources/SEGDatabaseManager.m

```diff
@@ -0,0 +1,419 @@
+static NSString *const kMainTableColumnCustomInstallationIdentifier =
+    @"custom_installation_identifier";
+static NSString *const kMainTableColumnFirebaseInstallationIdentifier =
+    @"firebase_installation_identifier";
+static NSString *const kMainTableColumnAssociationStatus = @"association_status";
+
+// Exclude the database from iCloud backup.
+static BOOL SegmentationAddSkipBackupAttributeToItemAtPath(NSString *filePathString) {
+  NSURL *URL = [NSURL fileURLWithPath:filePathString];
+  assert([[NSFileManager defaultManager] fileExistsAtPath:[URL path]]);
+
+  NSError *error = nil;
+  BOOL success = [URL setResourceValue:[NSNumber numberWithBool:YES]
+                                forKey:NSURLIsExcludedFromBackupKey
+                                 error:&error];
+  if (!success) {
+    // TODO(dmandar): log error.
+    NSLog(@"Error excluding %@ from backup %@.", [URL lastPathComponent], error);
+  }
+  return success;
+}
+
+static BOOL SegmentationCreateFilePathIfNotExist(NSString *filePath) {
+  if (!filePath || !filePath.length) {
+    // TODO(dmandar) log error.
+    NSLog(@"Failed to create subdirectory for an empty file path.");
+    return NO;
+  }
+  NSFileManager *fileManager = [NSFileManager defaultManager];
+  if (![fileManager fileExistsAtPath:filePath]) {
+    NSError *error;
+    [fileManager createDirectoryAtPath:[filePath stringByDeletingLastPathComponent]
+           withIntermediateDirectories:YES
+                            attributes:nil
+                                 error:&error];
+    if (error) {
+      // TODO(dmandar) log error.
+      NSLog(@"Failed to create subdirectory for database file: %@.", error);
+      return NO;
+    }
+  }
+  return YES;
+}
+
+@interface SEGDatabaseManager () {
+  /// Database storing all the config information.
+  sqlite3 *_database;
+  /// Serial queue for database read/write operations.
+  dispatch_queue_t _databaseOperationQueue;
+}
+@end
+
+@implementation SEGDatabaseManager
+
++ (instancetype)sharedInstance {
+  static dispatch_once_t onceToken;
+  static SEGDatabaseManager *sharedInstance;
+  dispatch_once(&onceToken, ^{
+    sharedInstance = [[SEGDatabaseManager alloc] init];
+  });
+  return sharedInstance;
+}
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    _databaseOperationQueue =
+        dispatch_queue_create("com.google.firebasesegmentation.database", DISPATCH_QUEUE_SERIAL);
+  }
+  return self;
+}
+
+#pragma mark - Public Methods
+
+- (void)loadMainTableWithCompletion:(SEGRequestCompletion)completionHandler {
+  __weak SEGDatabaseManager *weakSelf = self;
+  dispatch_async(_databaseOperationQueue, ^{
+    SEGDatabaseManager *strongSelf = weakSelf;
+    if (!strongSelf) {
+      completionHandler(NO, @{@"Database Error" : @"Internal database error"});
+    }
+
+    // Read the database into memory.
+    NSDictionary<NSString *, NSDictionary<NSString *, NSString *> *> *associations =
+        [self loadMainTable];
```

Comment:
`[self loadMainTable]` may return nil, which should be checked and treated as an error case here.